## Boas-vindas
Desejamos boas-vindas ao Google Colab! Uma plataforma que vai te auxiliar nesse primeiro passo que você está dando dentro da carreira em Ciência de Dados. Abaixo preparamos um breve resumo sobre esta ferramenta e suas funcionalidades.

## Introdução

O Google Colab é uma plataforma online gratuita que permite aos usuários executarem código Python em um ambiente de notebook Jupyter. **O que é notebook Jupyter?** É importante destacar que os códigos que são rodados no Colab não serão rodados em seu computador, mas sim em servidores da Google. Esse fato tem como consequência os principais motivos por termos escolhido o Google Colab para uso: permitir que os usuários colaborem em projetos com outras pessoas em tempo real, fazer com que os códigos dentro do Colab sejam executados sem ter que instalar nada no computador e, consequentemente, não causar o sobrecarregamento da sua máquina por causa da execução dos códigos (pois os códigos de manipulação de dados podem ser muito custosos computacionalmente).

## Integração com outros Serviços do Google

O Google Colab se integra a vários outros serviços do Google, incluindo o Google Drive e o Google Sheets (Google Planilhas). Isso torna fácil importar e exportar dados para e de notebooks do Colab. Os usuários também podem usar o Google Sheets como uma fonte de dados para modelos de aprendizado de máquina, o que pode ser especialmente útil para projetos que exigem dados em tempo real.

## Botando a mão na massa

Tá bom, chega de lero-lero. Provavelmente você não vê a hora de começar, e para isto reunimos alguns materiais importantes para te guiar neste inicio de jornada. Mas antes disso, aqui vai uma breve lista do que vamos abordar:
- Onde encontrar uma base de dados para utilizar;
- Como importar uma base de dados para o Google Colab;
- Como Manipular a base de dados importada.


# Introdução




Antes de prosseguirmos, é importante ressaltar que na comunidade de dados geralmente nos referimos a um "conjunto de dados" como "Dataset", então, a partir de agora usaremos apenas o termo Dataset, belê?

### Prosseguindo

Um local que recomendamos para encontrar Datasets, é o [Kaggle](https://kaggle.com/). Nele, uma vez que você encontrar o Dataset que quiser utilizar, você deve importá-lo para o Google Colab.
Para esse Colab vamos utilizar [winemag-data-130k-v2](https://www.kaggle.com/datasets/zynicide/wine-reviews). Para te ajudar nesse processo preparamos esse [tutorial](https://app.tango.us/app/workflow/Como-baixar-e-importar-um-Dataset-para-o-Colab-a7fa2f3515be4beab9c01197079dac3c) que mostra como baixar e ler o Dataset dentro do Colab.

O tutorial foi feito em uma versão anterior do Google colab. A unica distinção para o que temos agora é que na hora de dar permissões de acesso ao google drive, é necessário marcar todas as opções para permitir o acesso do Colab ao seu google drive pessoal. Na antiga versão, essas permissões eram concedidas de forma direta.

# Encontrando e importando um Dataset

Isso abaixo é uma célula de código, você pode clicar no botão de "play" para executar ela.
O código dentro dela irá "montar" o seu drive pessoal dentro do Colab.

*É necessário dar permissão para o colab acessar o seu Drive.*

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


A célula abaixo permitirá que o colab leia os conteúdos do nosso Dataset. Para que isso ocorra é necessário importar a biblioteca pandas

o import no python pode ser feito de algumas formas

1) import simples da biblioteca:
```
import pandas
```
2) apelidando a biblioteca:

```
import pandas as pd
```

nesse caso, quando chamamos a biblioteca pandas podemos escrever apenas 'pd' em vez de 'pandas'.

3) A última forma é importar apenas um módulo da biblioteca:

```
from pandas import DataFrame
```

aqui estamos importando apenas o módulo "DataFrame" da biblioteca "pandas"

Nesse nosso exemplo, usaremos a segunda forma.




In [ ]:
import pandas as pd

Uma vez importada a biblioteca, podemos utilizar a função read_csv() para ler o arquivo do google drive

OBS: verifique se o caminho passado dentro da função está correto em relação ao local que o arquivo foi salvo dentro do seu drive.

In [ ]:
drivePathCsv = "/content/drive/MyDrive/Colab Notebooks/USP/hype/Dojo_de_Machine_Learning/Módulo 1/winemag-data-130k-v2.csv"
df = pd.read_csv(drivePathCsv)

Agora que conseguimos ler o Dataset usando o Pandas, ele passa a ser um Dataframe. Por isso armazenamos ele numa variável chamada "df", mas poderia ser qualquer nome.

## sobre o Dataframe


Esse conjunto de dados que estamos utilizando como exemplo possui informações de vários vinhos e uma descrição feita por algum sommelier. As variáveis que ele possui são:

**country**: o país de origem do vinho

**description**: descrição do sommellier

**designation**: vinha de origem do vinho

**points**: nota que o vinho recebeu no "WineEnthusiast"

**price**: preço de uma garrafa do vinho

**province**: província (ou estado) de onde o vinho veio

**region_1**: província do plantio da uva

**region_2**: área específica dentro da província de plantio

**taster_name**: nome do sommelier

**taster_twitter_handle**: twitter do sommelier

**title**: nome do vinho

**variety**: tipo do vinho

**winery**: vinícula de origem do vinho

# Manipulando o Dataframe

Tendo o Dataframe em mãos e sabemos do que se trata, podemos prossguir para a etapa de manipulação.

## Visualizando trechos do Dataframe

Para verficar as primeiras 5 linhas do Dataframe, podemos utilizar a função head() do Pandas. Execute a célula abaixo para ver a saída.

In [ ]:
df.head()

Unnamed: 0   country                                        description  \
0           0     Italy  Aromas include tropical fruit, broom, brimston...   
1           1  Portugal  This is ripe and fruity, a wine that is smooth...   
2           2        US  Tart and snappy, the flavors of lime flesh and...   
3           3        US  Pineapple rind, lemon pith and orange blossom ...   
4           4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery  
0     White Blend              Nicosia  
1  Portuguese Red  Quinta dos Avidagos  
2      Pinot Gris            Rainstorm  
3        Riesling           St. Julian  
4      Pinot Noir         Sweet Cheeks

Podemos inserir um número como parâmetros para quantas linhas queremos ver. No exemplo abaixo utilizamos o valor "10", sinta-se a vontade para mudar e testar com os valores que quiser.

In [ ]:
df.head(10)

Unnamed: 0   country                                        description  \
0           0     Italy  Aromas include tropical fruit, broom, brimston...   
1           1  Portugal  This is ripe and fruity, a wine that is smooth...   
2           2        US  Tart and snappy, the flavors of lime flesh and...   
3           3        US  Pineapple rind, lemon pith and orange blossom ...   
4           4        US  Much like the regular bottling from 2012, this...   
5           5     Spain  Blackberry and raspberry aromas show a typical...   
6           6     Italy  Here's a bright, informal red that opens with ...   
7           7    France  This dry and restrained wine offers spice in p...   
8           8   Germany  Savory dried thyme notes accent sunnier flavor...   
9           9    France  This has great depth of flavor with its fresh ...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   
5                        Ars In Vitro      87   15.0     Northern Spain   
6                             Belsito      87   16.0  Sicily & Sardinia   
7                                 NaN      87   24.0             Alsace   
8                               Shine      87   12.0        Rheinhessen   
9                         Les Natures      87   27.0             Alsace   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   
5              Navarra                NaN   Michael Schachner   
6             Vittoria                NaN       Kerin O’Keefe   
7               Alsace                NaN          Roger Voss   
8                  NaN                NaN  Anna Lee C. Iijima   
9               Alsace                NaN          Roger Voss   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   
5           @wineschach  Tandem 2011 Ars In Vitro Tempranillo-Merlot (N...   
6          @kerinokeefe   Terre di Giurfo 2013 Belsito Frappato (Vittoria)   
7            @vossroger              Trimbach 2012 Gewurztraminer (Alsace)   
8                   NaN  Heinz Eifel 2013 Shine Gewürztraminer (Rheinhe...   
9            @vossroger  Jean-Baptiste Adam 2012 Les Natures Pinot Gris...   

              variety               winery  
0         White Blend              Nicosia  
1      Portuguese Red  Quinta dos Avidagos  
2          Pinot Gris            Rainstorm  
3            Riesling           St. Julian  
4          Pinot Noir         Sweet Cheeks  
5  Tempranillo-Merlot               Tandem  
6            Frappato      Terre di Giurfo  
7      Gewürztraminer             Trimbach  
8      Gewürztraminer          Heinz Eifel  
9          Pinot Gris   Jean-Baptiste Adam

Analogamente, podemos visualizar as últimas linhas do Dataframe usando a função tail() do Pandas.

In [ ]:
df.tail()

Unnamed: 0  country  \
129966      129966  Germany   
129967      129967       US   
129968      129968   France   
129969      129969   France   
129970      129970   France   

                                              description  \
129966  Notes of honeysuckle and cantaloupe sweeten th...   
129967  Citation is given as much as a decade of bottl...   
129968  Well-drained gravel soil gives this wine its c...   
129969  A dry style of Pinot Gris, this is crisp with ...   
129970  Big, rich and off-dry, this is powered by inte...   

                                   designation  points  price province  \
129966  Brauneberger Juffer-Sonnenuhr Spätlese      90   28.0    Mosel   
129967                                     NaN      90   75.0   Oregon   
129968                                   Kritt      90   30.0   Alsace   
129969                                     NaN      90   32.0   Alsace   
129970           Lieu-dit Harth Cuvée Caroline      90   21.0   Alsace   

       region_1      region_2         taster_name taster_twitter_handle  \
129966      NaN           NaN  Anna Lee C. Iijima                   NaN   
129967   Oregon  Oregon Other        Paul Gregutt           @paulgwine    
129968   Alsace           NaN          Roger Voss            @vossroger   
129969   Alsace           NaN          Roger Voss            @vossroger   
129970   Alsace           NaN          Roger Voss            @vossroger   

                                                    title         variety  \
129966  Dr. H. Thanisch (Erben Müller-Burggraef) 2013 ...        Riesling   
129967                  Citation 2004 Pinot Noir (Oregon)      Pinot Noir   
129968  Domaine Gresser 2013 Kritt Gewurztraminer (Als...  Gewürztraminer   
129969      Domaine Marcel Deiss 2012 Pinot Gris (Alsace)      Pinot Gris   
129970  Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...  Gewürztraminer   

                                          winery  
129966  Dr. H. Thanisch (Erben Müller-Burggraef)  
129967                                  Citation  
129968                           Domaine Gresser  
129969                      Domaine Marcel Deiss  
129970                          Domaine Schoffit

O mesmo conceito de quantidade de linhas se aplica a essa função.

In [ ]:
df.tail(10)

Unnamed: 0  country  \
129961      129961    Italy   
129962      129962    Italy   
129963      129963   Israel   
129964      129964   France   
129965      129965   France   
129966      129966  Germany   
129967      129967       US   
129968      129968   France   
129969      129969   France   
129970      129970   France   

                                              description  \
129961  Intense aromas of wild cherry, baking spice, t...   
129962  Blackberry, cassis, grilled herb and toasted a...   
129963  A bouquet of black cherry, tart cranberry and ...   
129964  Initially quite muted, this wine slowly develo...   
129965  While it's rich, this beautiful dry wine also ...   
129966  Notes of honeysuckle and cantaloupe sweeten th...   
129967  Citation is given as much as a decade of bottl...   
129968  Well-drained gravel soil gives this wine its c...   
129969  A dry style of Pinot Gris, this is crisp with ...   
129970  Big, rich and off-dry, this is powered by inte...   

                                   designation  points  price  \
129961                                     NaN      90   30.0   
129962               Sàgana Tenuta San Giacomo      90   40.0   
129963                                Oak Aged      90   20.0   
129964            Domaine Saint-Rémy Herrenweg      90    NaN   
129965             Seppi Landmann Vallée Noble      90   28.0   
129966  Brauneberger Juffer-Sonnenuhr Spätlese      90   28.0   
129967                                     NaN      90   75.0   
129968                                   Kritt      90   30.0   
129969                                     NaN      90   32.0   
129970           Lieu-dit Harth Cuvée Caroline      90   21.0   

                 province region_1      region_2         taster_name  \
129961  Sicily & Sardinia  Sicilia           NaN       Kerin O’Keefe   
129962  Sicily & Sardinia  Sicilia           NaN       Kerin O’Keefe   
129963            Galilee      NaN           NaN       Mike DeSimone   
129964             Alsace   Alsace           NaN          Roger Voss   
129965             Alsace   Alsace           NaN          Roger Voss   
129966              Mosel      NaN           NaN  Anna Lee C. Iijima   
129967             Oregon   Oregon  Oregon Other        Paul Gregutt   
129968             Alsace   Alsace           NaN          Roger Voss   
129969             Alsace   Alsace           NaN          Roger Voss   
129970             Alsace   Alsace           NaN          Roger Voss   

       taster_twitter_handle  \
129961          @kerinokeefe   
129962          @kerinokeefe   
129963        @worldwineguys   
129964            @vossroger   
129965            @vossroger   
129966                   NaN   
129967           @paulgwine    
129968            @vossroger   
129969            @vossroger   
129970            @vossroger   

                                                    title             variety  \
129961                        COS 2013 Frappato (Sicilia)            Frappato   
129962  Cusumano 2012 Sàgana Tenuta San Giacomo Nero d...        Nero d'Avola   
129963  Dalton 2012 Oak Aged Cabernet Sauvignon (Galilee)  Cabernet Sauvignon   
129964  Domaine Ehrhart 2013 Domaine Saint-Rémy Herren...      Gewürztraminer   
129965  Domaine Rieflé-Landmann 2013 Seppi Landmann Va...          Pinot Gris   
129966  Dr. H. Thanisch (Erben Müller-Burggraef) 2013 ...            Riesling   
129967                  Citation 2004 Pinot Noir (Oregon)          Pinot Noir   
129968  Domaine Gresser 2013 Kritt Gewurztraminer (Als...      Gewürztraminer   
129969      Domaine Marcel Deiss 2012 Pinot Gris (Alsace)          Pinot Gris   
129970  Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...      Gewürztraminer   

                                          winery  
129961                                       COS  
129962                                  Cusumano  
129963                                    Dalton  
129964                           Dom

Caso você prefira, poderá utilizar a função sample() do Pandas que, como o nome em inglês diz, exibirá uma amostra aleatória do Dataframe.

In [ ]:
df.sample(5)

Unnamed: 0      country  \
58052        58052           US   
114645      114645      Austria   
109038      109038  New Zealand   
1350          1350        Italy   
36743        36743           US   

                                              description      designation  \
58052   This blend of Pinot Grigio, Tocai Friuliano an...           Bianco   
114645  Savory herbs like lovage and chervil and ripe ...              NaN   
109038  Smoky, meaty and spicy, this wine boasts plent...  Jill's Vineyard   
1350    This is a lean, compact expression of Falanghi...            Preta   
36743   This hits it right in the sweet spot for Chard...              446   

        points  price          province           region_1       region_2  \
58052       91   32.0        California  Santa Ynez Valley  Central Coast   
114645      89   23.0  Südoststeiermark                NaN            NaN   
109038      87   28.0     Central Otago                NaN            NaN   
1350        86   19.0    Southern Italy             Sannio            NaN   
36743       85   13.0        California           Monterey  Central Coast   

             taster_name taster_twitter_handle  \
58052      Matt Kettmann         @mattkettmann   
114645  Anne Krebiehl MW           @AnneInVino   
109038    Joe Czerwinski                @JoeCz   
1350                 NaN                   NaN   
36743                NaN                   NaN   

                                                    title        variety  \
58052     Cent'Anni 2012 Bianco White (Santa Ynez Valley)    White Blend   
114645   Neumeister 2013 Grauburgunder (Südoststeiermark)  Grauburgunder   
109038  Rowland 2000 Jill's Vineyard Pinot Noir (Centr...     Pinot Noir   
1350    Capolino Perlingieri 2011 Preta Falanghina (Sa...     Falanghina   
36743          Noble Vines 2011 446 Chardonnay (Monterey)     Chardonnay   

                      winery  
58052              Cent'Anni  
114645            Neumeister  
109038               Rowland  
1350    Capolino Perlingieri  
36743            Noble Vines

Para ter uma melhor noção das dimensões do Dataframe usamos o atributo shape. Nesse caso temos 14 colunas e 12.9971 linhas. Lembrando que nosso Dataframe basicamente é uma tabela com 14 colunas e 12.9971 linhas.

In [ ]:
df.shape

(129971, 14)

A função a seguir é a describe(). Com ela podemos ter vários tipos de métricas úteis para uma análise superficial dos dados numéricos do Dataframe.

Somente são utilizadas as variáveis numéricas do dataset, uma vez que a função describe retorna valores estatísticos. Os valores retornados são:










*   **count**: número de observações daquela variável
*   **mean**: valor médio dessa variável
*   **std**: desvio padrão da variável
*   **min**: valor mínimo registrado
*   **25%**: valor que separa o primeiro quartil
*   **50%**: valor que separa o segundo quartil
*   **75%**: valor que separa o terceiro quartil
*   **max**: valor máximo registrado






In [ ]:
df.describe()

Unnamed: 0         points          price
count  129971.000000  129971.000000  120975.000000
mean    64985.000000      88.447138      35.363389
std     37519.540256       3.039730      41.022218
min         0.000000      80.000000       4.000000
25%     32492.500000      86.000000      17.000000
50%     64985.000000      88.000000      25.000000
75%     97477.500000      91.000000      42.000000
max    129970.000000     100.000000    3300.000000

O atributo columns fornece uma lista de todas as colunas do Dataframe

In [ ]:
df.columns

Index(['Unnamed: 0', 'country', 'description', 'designation', 'points',
       'price', 'province', 'region_1', 'region_2', 'taster_name',
       'taster_twitter_handle', 'title', 'variety', 'winery'],
      dtype='object')

A função info() do Pandas permite extrair informações como: quantidade de colunas, quantidade de linhas não nulas e o tipo de dado de cada uma.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129971 entries, 0 to 129970
Data columns (total 14 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   Unnamed: 0             129971 non-null  int64  
 1   country                129908 non-null  object 
 2   description            129971 non-null  object 
 3   designation            92506 non-null   object 
 4   points                 129971 non-null  int64  
 5   price                  120975 non-null  float64
 6   province               129908 non-null  object 
 7   region_1               108724 non-null  object 
 8   region_2               50511 non-null   object 
 9   taster_name            103727 non-null  object 
 10  taster_twitter_handle  98758 non-null   object 
 11  title                  129971 non-null  object 
 12  variety                129970 non-null  object 
 13  winery                 129971 non-null  object 
dtypes: float64(1), int64(2), object(11)


# Seleção de trechos do Dataframe

Todos os métodos descritos acima foram aplicados no dataframe completo. É possível também fazer uma seleção de trechos espcíficos do dataset.

A primeira forma que podemos "fatiar" o nosso dataset é por **colunas**.Para fazer isso, basta fornecer o nome da coluna desejada, entre aspas, dentro dos colchetes, como mostrado no exemplo abaixo.

In [ ]:
df['country']

0            Italy
1         Portugal
2               US
3               US
4               US
            ...   
129966     Germany
129967          US
129968      France
129969      France
129970      France
Name: country, Length: 129971, dtype: object

Isso que está na célula acima é uma Serie.

Serie é um tipo de dado muito similar a lista. É um a sequência de valores de um mesmo tipo. No caso, temos uma Serie de strings(palavras) que representam o país de origem do vinho.

É possível converter uma Serie para uma lista utilizando o seguinte código:


```
lista = df['country'].to_list()
```

Da mesma forma, podemos converter uma lista em uma Serie assim:

```
serie = pd.Series(lista)
```

Por serem estruturas muito parecidas, existem métodos que podem ser aplicados nas duas.

##splice em listas
Um exemplo que nos interessa é o splice de listas.

dado uma lista [1,3,5,7,9,11,13], é possível selecionar intervalos dela com a sintaxe:

```
lista[primeiro_elemento : ultimo_elemento ]
```

olhe o exemplo abaixo

Obs: note que a primeiro elemento é incluido na lista, mas o último não. Tome cuidado para passar os valores certos na hora de selecionar elementos da lista

In [ ]:
lista = [1,3,5,7,9,11,13]  # lista completa
lista

In [ ]:
lista[0:5] # entre elementos 0 e 5 (elemento 5 não incuído)

In [ ]:
lista[2:7] # entre elementos 2 e 7 (elemento 7 não incuído)

Podemos passar apenas um valor dentro do colchete para selecionar apenas um indice da lista

In [ ]:
lista[5] # elemento indice 5

In [ ]:
lista[3] # elemento indice 3

É possível passar valores negativos também, dessa forma a contagem de elementos é feita de traz para a frente

In [ ]:
lista[-3] # 3º elemento contando do final ([1,3,5,7,9,11,13])

In [ ]:
lista[-5] # 5º elemento contando do final ([1,3,5,7,9,11,13])

In [ ]:
lista[1:-3] # do elemento indice 1 até o indice -3 (contagem de traz para a frente)

## Splice em Series

Da mesma forma que em listas, podemos fazer esse fatiamento em Series.

In [ ]:
serie = df['country']
serie

0            Italy
1         Portugal
2               US
3               US
4               US
            ...   
129966     Germany
129967          US
129968      France
129969      France
129970      France
Name: country, Length: 129971, dtype: object

essa série de países tem 129971 observações

Vamos pegar a observação de indice 500

In [ ]:
serie[500]  # observação 500

'Spain'

In [ ]:
serie[12005] # observação 12005

'France'

In [ ]:
serie[10:20] # da observação 10 até a 20

10           US
11       France
12           US
13        Italy
14           US
15      Germany
16    Argentina
17    Argentina
18        Spain
19           US
Name: country, dtype: object

In [ ]:
serie[1000:-1000] # eliminando os primeiros e os ultimos 1000 registros

1000           US
1001        Italy
1002           US
1003       France
1004       France
           ...   
128966         US
128967    Austria
128968         US
128969         US
128970         US
Name: country, Length: 127971, dtype: object

## Seleção de mais de uma coluna

O conceito de series é muito importante pois um Dataframe nada mais é do que um varias series juntas. Por isso é importante saber manipulá-las bem.

Quando quisermos selecionar mais de uma coluna de um dataset, a estratégia é criar um dataset novo com as colunas desejadas.

Para fazer isso precisamos passar os nomes das colunas, entre aspas, entre **DOIS** colchetes

In [ ]:
df_novo = df[['country','variety']]
df_novo

country         variety
0          Italy     White Blend
1       Portugal  Portuguese Red
2             US      Pinot Gris
3             US        Riesling
4             US      Pinot Noir
...          ...             ...
129966   Germany        Riesling
129967        US      Pinot Noir
129968    France  Gewürztraminer
129969    France      Pinot Gris
129970    France  Gewürztraminer

[129971 rows x 2 columns]

In [ ]:
df_novo2 = df[['country','variety','province', 'region_1',]]
df_novo2

country         variety           province             region_1
0          Italy     White Blend  Sicily & Sardinia                 Etna
1       Portugal  Portuguese Red              Douro                  NaN
2             US      Pinot Gris             Oregon    Willamette Valley
3             US        Riesling           Michigan  Lake Michigan Shore
4             US      Pinot Noir             Oregon    Willamette Valley
...          ...             ...                ...                  ...
129966   Germany        Riesling              Mosel                  NaN
129967        US      Pinot Noir             Oregon               Oregon
129968    France  Gewürztraminer             Alsace               Alsace
129969    France      Pinot Gris             Alsace               Alsace
129970    France  Gewürztraminer             Alsace               Alsace

[129971 rows x 4 columns]

## Seleção de linhas

###loc e iloc

Agora que voce já sabe separar colunas, vamos para a seleção de linhas.

Em séries individuais, podemos usar o método de slice normalmente. Mas, quando estamos trabalhando com um dataframe precisamos utilizar os métodos .loc e .iloc

Para os dois métodos, devemos utilizar a mesma sintaxe de splice em listas mas com eses métodos na frente.

Aqui abaixo tem um exemplo do uso.

In [ ]:
df_novo2.iloc[5] #linha 5

country                  Spain
variety     Tempranillo-Merlot
province        Northern Spain
region_1               Navarra
Name: 5, dtype: object

isso retorna os valores de todas as variáveis na linha

In [ ]:
df_novo2.iloc[5:20] # entre linha 5 e 20

country             variety           province          region_1
5       Spain  Tempranillo-Merlot     Northern Spain           Navarra
6       Italy            Frappato  Sicily & Sardinia          Vittoria
7      France      Gewürztraminer             Alsace            Alsace
8     Germany      Gewürztraminer        Rheinhessen               NaN
9      France          Pinot Gris             Alsace            Alsace
10         US  Cabernet Sauvignon         California       Napa Valley
11     France      Gewürztraminer             Alsace            Alsace
12         US  Cabernet Sauvignon         California  Alexander Valley
13      Italy   Nerello Mascalese  Sicily & Sardinia              Etna
14         US          Chardonnay         California     Central Coast
15    Germany            Riesling              Mosel               NaN
16  Argentina              Malbec              Other          Cafayate
17  Argentina              Malbec   Mendoza Province           Mendoza
18      Spain   Tempranillo Blend     Northern Spain  Ribera del Duero
19         US            Meritage           Virginia          Virginia

A única diferença entre .loc e .iloc é que com o iloc é passado a posição numérica da linha desejada.

Já no .loc devemos passar o rótulo da linha e não a usa posição (o rótulo seria essa primeira coluna que fica a esquerda da tabela com os números em negrito)

Nesse caso, como as linhas estão ordenadas o loc e o iloc fazem a mesma coisa. Mas nos casos em que o rótulo não coincide com a posição devemos nos atentar a qual usar.

###Filtro por valor

# Outra forma que podemos utilizar para filtrar linhas desejadas do dataset é com a síntaxe abaixo



```
df[(condicao_desejada)]
```

Esse exemplo abaixo separa apenas os vinhos de origem francesa.


In [ ]:
df_novo = df[df['country'] == 'France']  #filtrando quando df['country'] for igual a 'France'
df_novo.head()

Unnamed: 0 country                                        description  \
7            7  France  This dry and restrained wine offers spice in p...   
9            9  France  This has great depth of flavor with its fresh ...   
11          11  France  This is a dry wine, very spicy, with a tight, ...   
30          30  France  Red cherry fruit comes laced with light tannin...   
42          42  France  This is a festive wine, with soft, ripe fruit ...   

    designation  points  price    province             region_1 region_2  \
7           NaN      87   24.0      Alsace               Alsace      NaN   
9   Les Natures      87   27.0      Alsace               Alsace      NaN   
11          NaN      87   30.0      Alsace               Alsace      NaN   
30      Nouveau      86    NaN  Beaujolais  Beaujolais-Villages      NaN   
42      Nouveau      86    9.0  Beaujolais           Beaujolais      NaN   

   taster_name taster_twitter_handle  \
7   Roger Voss            @vossroger   
9   Roger Voss            @vossroger   
11  Roger Voss            @vossroger   
30  Roger Voss            @vossroger   
42  Roger Voss            @vossroger   

                                                title         variety  \
7               Trimbach 2012 Gewurztraminer (Alsace)  Gewürztraminer   
9   Jean-Baptiste Adam 2012 Les Natures Pinot Gris...      Pinot Gris   
11            Leon Beyer 2012 Gewurztraminer (Alsace)  Gewürztraminer   
30  Domaine de la Madone 2012 Nouveau  (Beaujolais...           Gamay   
42             Henry Fessy 2012 Nouveau  (Beaujolais)           Gamay   

                  winery  
7               Trimbach  
9     Jean-Baptiste Adam  
11            Leon Beyer  
30  Domaine de la Madone  
42           Henry Fessy

É possível passar mais de uma condição para o filtro também, basta usar os operadores lógicos "&" e "|" e colocar as condições entre parenteses

In [ ]:
df_novo = df[(df['country'] == 'France') & (df['designation'] == 'Nouveau')]  #filtrando quando df['country'] for igual a 'France' e df['designation'] == 'Nouveau'
df_novo.head()

Unnamed: 0 country                                        description  \
30            30  France  Red cherry fruit comes laced with light tannin...   
42            42  France  This is a festive wine, with soft, ripe fruit ...   
2431        2431  France  This has surprisingly structured fruits and fi...   
7646        7646  France  Henry Fessy's can-can dancer is with us again ...   
9291        9291  France  In keeping with the good quality of 2015's Bea...   

     designation  points  price    province             region_1 region_2  \
30       Nouveau      86    NaN  Beaujolais  Beaujolais-Villages      NaN   
42       Nouveau      86    9.0  Beaujolais           Beaujolais      NaN   
2431     Nouveau      85   11.0  Beaujolais           Beaujolais      NaN   
7646     Nouveau      85    9.0  Beaujolais           Beaujolais      NaN   
9291     Nouveau      86   13.0  Beaujolais  Beaujolais-Villages      NaN   

     taster_name taster_twitter_handle  \
30    Roger Voss            @vossroger   
42    Roger Voss            @vossroger   
2431  Roger Voss            @vossroger   
7646  Roger Voss            @vossroger   
9291  Roger Voss            @vossroger   

                                                  title variety  \
30    Domaine de la Madone 2012 Nouveau  (Beaujolais...   Gamay   
42               Henry Fessy 2012 Nouveau  (Beaujolais)   Gamay   
2431         Georges Duboeuf 2012 Nouveau  (Beaujolais)   Gamay   
7646             Henry Fessy 2011 Nouveau  (Beaujolais)   Gamay   
9291  Domaine de la Madone 2015 Nouveau  (Beaujolais...   Gamay   

                    winery  
30    Domaine de la Madone  
42             Henry Fessy  
2431       Georges Duboeuf  
7646           Henry Fessy  
9291  Domaine de la Madone

# Agrupamento por valor




Uma outra ferramenta muito interessante que temos para organizar dataframes que temos é a função groupby().

Esse é o tipo de função que é mais fácil ver funcionando do que explicar o que ela faz.

Passamos uma variável do dataset como parâmetro da função que será utilizada como referência para agrupar as outras variáveis numéricas(nesse caso "country"). Mas como assim agrupar as outras variáveis numéricas??

A forma de agrupamento é passada com esse método no final (.mean()) isso diz ao programa que o nosso objetivo é juntar todas as observações de um mesmo país e tirar a média os valores numéricos das outras variáveis.

Primeiro, vamos criar um dataset com apenas a coluna que queremos utilizar como base para o agrupamento e as colunas numéricas de interesse.

Após isso, aplicamos a função groupby no dataset

In [27]:
df_group = df[['country','points','price']]

df_group = df_group.groupby('country').mean()
df_group

points      price
country                                     
Argentina               86.710263  24.510117
Armenia                 87.500000  14.500000
Australia               88.580507  35.437663
Austria                 90.101345  30.762772
Bosnia and Herzegovina  86.500000  12.500000
Brazil                  84.673077  23.765957
Bulgaria                87.936170  14.645390
Canada                  89.369650  35.712598
Chile                   86.493515  20.786458
China                   89.000000  18.000000
Croatia                 87.219178  25.450704
Cyprus                  87.181818  16.272727
Czech Republic          87.250000  24.250000
Egypt                   84.000000        NaN
England                 91.581081  51.681159
France                  88.845109  41.139120
Georgia                 87.686047  19.321429
Germany                 89.851732  42.257547
Greece                  87.283262  22.364425
Hungary                 89.191781  40.648276
India                   90.222222  13.333333
Israel                  88.471287  31.768916
Italy                   88.562231  39.663770
Lebanon                 87.685714  30.685714
Luxembourg              88.666667  23.333333
Macedonia               86.833333  15.583333
Mexico                  85.257143  26.785714
Moldova                 87.203390  16.745763
Morocco                 88.571429  19.500000
New Zealand             88.303030  26.931785
Peru                    83.562500  18.062500
Portugal                88.250220  26.218256
Romania                 86.400000  15.241667
Serbia                  87.500000  24.500000
Slovakia                87.000000  16.000000
Slovenia                88.068966  24.812500
South Africa            88.056388  24.668987
Spain                   87.288337  28.215275
Switzerland             88.571429  85.285714
Turkey                  88.088889  24.633333
US                      88.563720  36.573464
Ukraine                 84.071429   9.214286
Uruguay                 86.752294  26.403670

Dessa forma, podemos ver qual a média de cada variável numérica para cada um dos países. É possível, por exemplo, ver a média de preço em cada país e saber quais são os países que possuem o vinho mais caro ou o mais barato.

OBS: o ".mean()" poderia também ser: .sum(), .std(), .min() e .max() (soma dos valores, desvio padrão, valor mínimo e valor máximo respectivamente)

Caso queira ordenar os valores, podemos utilzar a função

```
df_group.sort_values(by='points', ascending=True)
```
devemos passar o nome da coluna considerada na ordenação no atributo "by". O atributo ascending diz se será ordenado de forma crescente ou decrescente. Ascending = True é ordenação crescente e ascending = False é ordenação decrescente

vej como fica os países ordenados pela média de pontuação por exemplo

In [ ]:
df_group.sort_values(by='points', ascending=False)

# Dados nulos

É importante que você saiba que algo muito comum ao se trabalhar com datasets é a existência de dados faltantes, também conhecidos como dados nulos. Dados faltantes basicamente são dados não registrados, ou seja, uma determinada célula do dataframe que não possui nenhum dado lá dentro. Saiba que é comum que os dados nulos sejam chamados de "NaN" (Not a Number) no dataframe. Por fim, é importante que você saiba quais motivos levam a um dado faltante.

## Dados nulos aleatórios

O primeiro motivo para a existência para um dado nulo é por causa aleatória. Daremos alguns exemplos para que fique mais claro:



*   Imagine que há um dispositivo que mede a salinidade da água. Esse dispositivo consiste em uma boia com um cilindro metálico acoplado. A boia é jogada no mar, e tem como finalidade fazer com que o dispositivo como um todo flutue. Por sua vez, o cilindro metálico é a parte que justamente tem como finalidade medir a salinidade da água. O cilindro mede e armazena a salinidade que ele detectou na água a cada exatos 1 minuto. Entretanto, devido a agitação das ondas do mar, a boia pode se mexer de forma que, quando o cilindro for medir a salinidade, o cilindro se encontra fora da água. E quando isso acontece, o dispositivo armazena um dado nulo, pois não foi possível armazenar um dado de salinidade da água. Perceba que nesse caso o motivo da existência do dado nulo é totalmente aleatória. Ou seja, não há um padrão para a existência do dado nulo.
*   Um questionário pode ser perdido acidentalmente no correio de forma aleatória (imprevisível)
*   Uma amostra de sangue pode ser danificada acidentalmente no laboratório de forma aleatória (imprevisível)

Portanto, quando dizemos que faltam dados completamente ao acaso, queremos dizer que a falta não tem nada a ver com o objeto que está sendo estudado.







## Dados nulos devido a razão sistemática


Por outro lado, também há a possibilidade de que a existência dos dados nulos não seja aleatória. Ou seja, pode ser que a falta se deva a uma razão sistemática. Vamos para alguns exemplos para que isso fique mais claro:


*   Imagine que está sendo feita uma pesquisa, sendo que essa pesquisa é feita de forma a perguntar ao entrevistado algumas informações sobre si. Porém, o candidato pode optar por não querer responder a uma determinada pergunta. Uma das perguntas é declarar qual é a renda mensal do entrevistado. É conhecido o fato de que indivíduos com alta renda tendem a não revelar o quanto ganham de salário (para fins didáticos, imagine que pessoas com baixa ou média renda tendem a declarar sem problemas seu salário). Quando os dados de todos os entrevistados foram coletados, percebeu-se que alguns candidatos não declararam qual era seu salário, ou seja, possuíam um dado nulo. Porém, perceba que neste caso o dado nulo não se deve à causa aleatória, pois sabemos que eles se devem provavelmente ao fato daquela pessoa possui alta renda. Neste caso, a existência do dado nulo está provavelmente atrelada a um motivo conhecido.
*   Imagine que uma pessoa deveria comparecer hoje a um exame para detectar se ela está sob o efeito de drogas. Entretanto, a pessoa não compareceu ao exame (e também não será punida por isso). Ou seja, o resultado daquele exame se tornou um dado nulo, uma informação faltante em relação àquela pessoa. Porém, sabe-se que uma pessoa que não está drogada não deveria ter problemas em fazer o exame, ou seja, se a pessoa faltou no exame é devido uma tentativa de esconder que está drogada. Perceba que novamente neste caso o motivo da existência do dado nulo não é totalmente aleatória, pois a existência do dado nulo está provavelmente atrelada a um motivo conhecido (sistemático).


Perceba que nesse contexto, diferentemente dos dados nulos aleatórios, a existência de um dado nulo nos revela uma (provável) informação sobre aquela pessoa sendo estudada. Ou seja, o dado nulo nos fornece uma informação sobre o objeto estudado.





# Como identificar e tratar dados nulos

Nesta seção você irá estudar sobre como identificar dados nulos em um dataframe, assim como de que forma pode tratá-los. Inicialmente, comece estudando os conteúdos abaixo:

[Como lidar com dados faltantes (NaN) em um Dataset](https://www.youtube.com/watch?v=k1zi4EwIXoc)

[Como tratar dados nulos no dataset?](https://medium.com/horadecodar/como-tratar-dados-nulos-no-dataset-4f0470b22d38)

A seguir, revisaremos os conteúdos abordados!

A função isnull() mostrará se cada linha possui algum valor nulo.

In [ ]:
df.isnull()

A função isnull() por si só não nos fornece uma visão muito boa, pois seria necessário vasculhar o Dataset inteiro procurando por algum "True".

Para isso pode ser melhor usar a função sum(), que vai somar e retornar a quantidade de valores nulos para cada coluna.

In [ ]:
df.isnull().sum()

Como visto acima, existem muitos valores nulos e isso pode nos atrapalhar.

Uma forma de tratar esse problema é eliminando todas as linhas com valores nulos.

Para isso existe a função dropna() do Pandas. Ela remove todas as linhas que possuem algum valor nulo.

In [ ]:
df = df.dropna()
df

# Dados duplicados

Para verificar linhas duplicadas usamos a função duplicated(). Porém, ela possui o mesmo problema da função isnull(), isto é, ela é dificil de interpretar.

In [28]:
df.duplicated()

0         False
1         False
2         False
3         False
4         False
          ...  
129966    False
129967    False
129968    False
129969    False
129970    False
Length: 129971, dtype: bool

Para facilitar, podemos usar novamente a função sum(), que nos retornará a quantidade de linhas duplicadas. No caso não existem linhas duplicadas.

In [ ]:
df.duplicated().sum()

Mas se existissem, poderíamos usar a função drop_duplicates().

In [ ]:
df = df.drop_duplicates()

# E aí curtiu?

Agora é sua vez! Vá no Colab chamado "Desafio" e complete as informações faltantes;